weave/index/index_sql.py

```diff
@@ -0,0 +1,869 @@
+        """Close the connection to the database."""
+        self.cur.close()
+        self.con.close()
+
+    def _connect(self, database_name="weave_db", **kwargs):
+        """Connect to the server, and select the specified DB.
+
+        Parameters
+        ----------
+        database_name: str (default='weave_db')
+            The database to connect to.
+        **driver: str (default='{ODBC Driver 18 for SQL Server}')
+            Driver to be used for the connection. Usually similar to:
+            '{ODBC Driver <18, 17, 13...> for SQL Server}'.
+        **encrypt: bool (default=False)
+            Whether or not to encrypt the database (disabled by default, due
+            to the fact that it requires additional configuration).
+        """
+        driver = kwargs.get("odbc_driver", "{ODBC Driver 18 for SQL Server}")
+        encrypt = kwargs.get("encrypt", False)
+
+        # Pylint has a problem recognizing 'connect' as a valid member function
+        # so we ignore that here.
+        # pylint: disable-next=c-extension-no-member
+        self.con = pyodbc.connect(
+            f"DRIVER={driver};"
+            f"SERVER={os.environ['MSSQL_HOST']};"
+            f"DATABASE={database_name};"
+            f"UID={os.environ['MSSQL_USERNAME']};"
+            f"PWD={os.environ['MSSQL_PASSWORD']};"
+            f"Encrypt={'yes' if encrypt else 'no'};"
+        )
+        self.cur = self.con.cursor()
+
+    def _select_db(self, database_name="weave_db"):
+        """Select the database to use, or create it if it does not exist.
+
+        Parameters
+        ----------
+        database_name: str (default='weave_db')
+            The database to connect to.
+        """
+        # Query to see if the database exists.
+        self.cur.execute(
+            "SELECT * FROM sys.databases WHERE name = ?", (database_name,)
+        )
+        # If the database does not exist, create it.
+        if not self.cur.fetchone():
+            self.con.autocommit = True
+            self.cur.execute(f"CREATE DATABASE {database_name};")
+            self.cur.execute(f"USE {database_name};")
+            self.con.autocommit = False
+        else:
+            self.cur.execute(f"USE {database_name};")
+
+    def _create_schema(self):
+        """Create the schema if it does not already exist."""
+        # Check if self.pantry_schema exists.
+        self.cur.execute(
+            "SELECT * FROM sys.schemas WHERE name = ?", (self.pantry_schema,)
+        )
+        # If it does not exist, create it.
+        if not self.cur.fetchone():
+            self.cur.execute(f"CREATE SCHEMA {self.pantry_schema};")
+            self.con.commit()
+
+    def _create_tables(self):
+        """Create the required tables if they do not already exist."""
+        # THIS NEEDS TO BE UPDATED MANUALLY IF NEW COLUMNS ARE ADDED TO INDEX.
+        # THE INSERT IN OTHER FUNCTIONS USE config.index_schema(), BUT THAT
+        # CAN'T BE USED HERE AS TYPE NEEDS TO BE SPECIFIED.
```

Comment:
WHY ARE YOU YELLING 😱

Reply:
:skull_and_crossbones: BECAUSE ITS IMPORTANT! :joy: